perun-core/src/main/java/cz/metacentrum/perun/core/impl/ConsentsManagerImpl.java

```diff
@@ -51,11 +55,303 @@
 		return consentHub;
 	};
 
+	// Consent mapper
+	private final static RowMapper<Consent> CONSENT_MAPPER = (rs, rowNum) -> {
+		Consent c = new Consent();
+		c.setId(rs.getInt("consents_id"));
+		c.setUserId(rs.getInt("consents_user_id"));
+		c.setConsentHub(CONSENT_HUB_MAPPER.mapRow(rs, rowNum));
+		c.setStatus(ConsentStatus.valueOf(rs.getString("consents_status")));
+		c.setCreatedAt(rs.getString("consents_created_at"));
+		c.setCreatedBy(rs.getString("consents_created_by"));
+		c.setModifiedAt(rs.getString("consents_modified_at"));
+		c.setModifiedBy(rs.getString("consents_modified_by"));
+		if(rs.getInt("consents_modified_by_uid") == 0) c.setModifiedByUid(null);
+		else c.setModifiedByUid(rs.getInt("consents_modified_by_uid"));
+		if(rs.getInt("consents_created_by_uid") == 0) c.setCreatedByUid(null);
+		else c.setCreatedByUid(rs.getInt("consents_created_by_uid"));
+		return c;
+	};
+
+	private final static RowMapper<Consent> CONSENT_MAPPER_WITHOUT_HUB = (rs, rowNum) -> {
+		Consent c = new Consent();
+		c.setId(rs.getInt("consents_id"));
+		c.setUserId(rs.getInt("consents_user_id"));
+		c.setConsentHub(new ConsentHub());
+		c.setStatus(ConsentStatus.valueOf(rs.getString("consents_status")));
+		c.setCreatedAt(rs.getString("consents_created_at"));
+		c.setCreatedBy(rs.getString("consents_created_by"));
+		c.setModifiedAt(rs.getString("consents_modified_at"));
+		c.setModifiedBy(rs.getString("consents_modified_by"));
+		if(rs.getInt("consents_modified_by_uid") == 0) c.setModifiedByUid(null);
+		else c.setModifiedByUid(rs.getInt("consents_modified_by_uid"));
+		if(rs.getInt("consents_created_by_uid") == 0) c.setCreatedByUid(null);
+		else c.setCreatedByUid(rs.getInt("consents_created_by_uid"));
+		return c;
+	};
+
+
+	private static JdbcPerunTemplate jdbc;
+
+	protected final static String consentHubMappingSelectQuery = "consent_hubs.id as consent_hubs_id, consent_hubs.name as consent_hubs_name, consent_hubs.enforce_consents as consent_hubs_enforce_consents, " +
+		"consent_hubs.created_at as consent_hubs_created_at, consent_hubs.created_by as consent_hubs_created_by, consent_hubs.modified_at as consent_hubs_modified_at, " +
+		"consent_hubs.modified_by as consent_hubs_modified_by, consent_hubs.created_by_uid as consent_hubs_created_by_uid, consent_hubs.modified_by_uid as consent_hubs_modified_by_uid";
+
+
 	public ConsentsManagerImpl(DataSource perunPool) {
 		jdbc = new JdbcPerunTemplate(perunPool);
 		jdbc.setQueryTimeout(BeansUtils.getCoreConfig().getQueryTimeout());
 	}
 
+	@Override
+	public Consent createConsent(PerunSession perunSession, Consent consent) throws ConsentNotExistsException, ConsentExistsException, ConsentHubNotExistsException {
+		// Check if consent already exists
+		if(consentExists(perunSession, consent)){
+			throw new ConsentExistsException("Consent already exists.");
+		}
+
+		// Check if UNSIGNED consent exists and delete it
```

Comment:
If you move the check in BL from deleteConsent to createConsent, this is no longer needed (On second thought, the check also probably **should** be in BL, not here.)